### PR TITLE
Repackages code for pypi

### PR DESCRIPTION
**Pull request by oke-aditya**
Updates for better PyPi packaging. Helps #182. #209

So now how the installation goes.
Note that

If we are at the repo root. then we do `pip install .` otherwise on PyPi we would just replace `.` with `mantisshrimp`

With this update, we will be able to install engines separately through `setup.py` itself. This follows the HuggingFace strategy exactly. Which we aimed for.
 
- To install mantisshrimp raw. without any engine
`pip install mantisshrimp` or if we have cloned and we are installing `pip install .`

- To install with Fastai.
`pip install mantisshrimp[fastai]` or if we have cloned and we are installing `pip install .[fastai]`

- To install with Pytorch-lightning
`pip install mantisshrimp[pytorch-lightning]` or if we have cloned and we are installing `pip install .[pytorch-lightning]`

To install completely with both the engines
`pip install mantisshrimp[all]` or if we have cloned and we are installing `pip install .[all]`

Note that we cannot provide the all option as default as these packages are built incrementally and that's what even HF follows.

We expect the user to have torch and torchvision pre-installed otherwise we do that for them.

I see our CI is kinda broke. It installs from `requirements.txt` and `requirements-extra.txt`. This will not work henceforth. If we are packaging this code then we should have `pip install .[all]` in our repo base and check for the tests. `requirements.txt` and `requirements-extra.txt` can completely be deleted then.

This CI change will ensure that what we build in `setup.py` is what we test and will be what we will deploy. Also it will ensure that correct requirements are listed in `settings.ini` and there is no discrepancy.

I will raise PR for CI change. Then re-run this to check for it.

This current installation method. I have checked locally all seem to be working good. Also, I will add a CI to deploy to PyPi when we tag a release later.

### CODE REVIEW

File: settings.ini [hunk 1]
```diff
@@ -25,7 +25,9 @@ license = apache2
 status = 2
 
 # Optional. Same format as setuptools requirements
-requirements = fastcore torch torchvision
+requirements = fastcore tqdm matplotlib torch torchvision
```
Comment (lgvaz) [inline settings.ini:28]: Can we take out torch and torchvision from here and specify it should be previously installed?

**Comment by lgvaz** [comment]
> Note that we can provide the all option as default

That seems better, can we go with that? 

And then have a [minimal] tag

**Comment by lgvaz** [comment]
There are some missing packages to install, for example, fastai requires `ipykernel`

We also need opencv

Also, how can we handle other additional deps like efficientdet?

---
A request, before we merge this, can you start from an empty env and confirm that we can run `getting-started` with faster_rcnn and efficientdet with both training engines?

You should actually create one env for each engine and confirm they are working individually

**Comment by lgvaz** [comment]
All should also contain additional packages like `albumentations`

File: settings.ini (hunk 1 above)
Comment (oke-aditya) [reply settings.ini:28]: We can take that out. But we need to ensure that the user has these two installed. If they aren't then how would we handle?

**Comment by oke-aditya** [comment]
> 
> 
> > Note that we can provide the all option as default
> 
> That seems better, can we go with that?
> 
> And then have a [minimal] tag

I did a typo. Note that we **cannot** provide the all option as default. So by default the installation is Minimal.

**Comment by oke-aditya** [comment]
> 
> 
> There are some missing packages to install, for example, fastai requires `ipykernel`
> 
> We also need opencv
> 
> Also, how can we handle other additional deps like efficientdet?
> 
> A request, before we merge this, can you start from an empty env and confirm that we can run `getting-started` with faster_rcnn and efficientdet with both training engines?
> 
> You should actually create one env for each engine and confirm they are working individually

All these sub-requirements like `ipykernel` would be installed by fastai2 itself. For efficientdet and other repos I think the default installation should come with all of these. Even when deploying you need to instantiate the models.

File: settings.ini (hunk 1 above)
Comment (oke-aditya) [reply settings.ini:28]: While testing Torch triggered a wrong package, pytorch should be done by the user. We will automate fastai and lightning. Agreed and changed. Torch and torhvision go hand in hand so better user handle from PyTorch.

**Comment by oke-aditya** [comment]
I am testing stuff locally. Also I will edit the current CI so that stuff works perfectly.

**Comment by oke-aditya** [comment]
> 
> 
> All should also contain additional packages like `albumentations`, `pytest`

Is albumentations mandatory for user ? He can use out of the box torchvision transforms too right?

**Comment by oke-aditya** [comment]
We need a fix for #181 Without this it is becoming very hard to limit the dependencies. Everything is imported everywhere so packaging might not work well. Also I need to write seperate tests for users using only lightning and users using only fastai engine.
We need to refactor the tests folder into these two subfolders. Then we can do proper Testing.


As of now I am updating the settings.ini with suggestions. But we need the above stuff.

**Comment by oke-aditya** [comment]
As of now I have passed the tests with this update of dependencies. I will add a PR for CI

**Comment by oke-aditya** [comment]
Yeah, so this builds with the new CI test itself. So I guess with this we will have both stuff done.

**Comment by oke-aditya** [comment]
We need revamp of imports in tests folder as well. It is again importing everything unnecessary.
I will make sure that this packaging works well for now. This will help us while we revamp the unnecessary imports. 
In a later PR I will try to Prune the dependencies. For now this seems easy and feasible solution. I do not want to add many changes and keep breaking CI or requirements.

**Comment by oke-aditya** [comment]
CI Test fails because of Ross efficient Det. I am unable add this to `setup.py` as it is a git repo. `setup.py` expects a egg file or pypi package. For now I am by passing this by adding it to CI testing file itself as bash command.

We need to fix this, if some-one can let me know how to package such github repos in `settings.ini` or in `setup.py` it would be easy fix.

Otherwise, we need help from Ross or build an egg file from his repo and ship it.

**Comment by oke-aditya** [comment]
All CI pass. I have patched and fixed stuff.

Issues that still pertain: -
1. Separate CI for engines. This is CI for All engines and complete installation (some extra unnecessary stuff too). And a seperate CI for no engine deployment ready mantisshrimp. We need to mark our tests in order to support these.
2. How to add Ross Efficient det to setup.py. It is not getting bundled. I have patched it in CI which is not correct.
3. CI action to deploy this package with tag push.
4. Fix up the imports !!! they will help in Pruning Packages. Maybe some we do not need in minimal installation.

**Comment by oke-aditya** [comment]
Thoughts @paras-jain @ai-fast-track @lgvaz

File: .github/workflows/ci-all-testing.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # This workflow will install Python dependencies, run tests and lint with a variety of Python versions
 # For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
 
-name: tests
+name: CI All Test
```
Comment (lgvaz) [inline .github/workflows/ci-all-testing.yml:4]: just call it tests for now

**Comment by lgvaz** [comment]
Btw, do we currently have the default `pip install mantisshrimp` that installs everything?

**Comment by oke-aditya** [comment]
No, but we have option `pip install mantisshrimp[all]` that installs everything. We cannot make `pip install mantisshrimp` to install all

File: .github/workflows/ci-all-testing.yml (hunk 2 above)
Comment (oke-aditya) [reply .github/workflows/ci-all-testing.yml:4]: K renamed 👍  just keeping stuff as `Tests` Title looks easier on the eyes.

**Pull request merged**